lib/streamlit/navigation/page.py

```diff
@@ -0,0 +1,159 @@
+    >>> import streamlit as st
+    >>>
+    >>> def page2():
+    >>>     st.title("Second page")
+    >>> pg = st.navigation([
+    >>>	    st.Page("page1.py", title="First page", icon="🔥"),
+    >>>	    st.Page(page2, title="Second page", icon=":material/favorite:"),
+    >>> ])
+    >>> pg.run()
+    """
+    return StreamlitPage(page, title=title, icon=icon, default=default)
+
+
+class StreamlitPage:
+    def __init__(
+        self,
+        page: str | Path | Callable[[], None],
+        *,
+        title: str | None = None,
+        icon: str | None = None,
+        default: bool = False,
+    ):
+        ctx = get_script_run_ctx()
+        if not ctx:
+            return
+
+        main_path = Path(ctx.pages_manager.main_script_path).parent
+        if isinstance(page, str):
+            page = Path(page)
+        if isinstance(page, Path):
+            page = (main_path / page).resolve()
+
+        inferred_name = ""
+        inferred_icon = ""
+        if isinstance(page, Path):
+            inferred_icon, inferred_name = page_icon_and_name(page)
+        elif title is None:
+            if hasattr(page, "__name__"):
+                inferred_name = str(page.__name__)
+            else:
+                raise StreamlitAPIException(
+                    "Cannot infer page title for Callable. Set the `title=` keyword argument."
+                )
+
+        self._page: Path | Callable[[], None] = page
+        self.title: str = title or inferred_name
+        self.icon: str = icon or inferred_icon
+
+        if self.icon:
+            validate_icon_or_emoji(self.icon)
+
+        self.default: bool = default
+        # used by st.navigation to ordain a page as runnable
+        self._can_be_called: bool = False
+
+    def run(self) -> None:
+        if not self._can_be_called:
+            raise StreamlitAPIException(
+                "This page cannot be called directly. Only the page returned from st.navigation can be called once."
+            )
+
+        self._can_be_called = False
+
+        ctx = get_script_run_ctx()
+        if not ctx:
+            return
+
+        ex = None
+        with ctx.pages_manager.run_with_active_hash(self._script_hash):
+            try:
+                if callable(self._page):
+                    self._page()
+                    return
+                else:
+                    code = ctx.pages_manager.get_page_script_byte_code(str(self._page))
+
+                    module = types.ModuleType("__page__")
+                    module.__dict__["__file__"] = self._page
+                    exec(code, module.__dict__)
+            except Exception as e:
+                # Catch the exception so we can raise it after we've reset the active page
+                ex = e
+
+        if ex:
+            raise ex
+
+    @property
+    def _script_hash(self) -> str:
+        if isinstance(self._page, Path):
+            h = calc_md5(str(self._page))
+        else:
+            h = calc_md5(self.title)
```

Comment:
if we calc the hash based on the title, we should perhaps add a check in `navigation` for title uniqueness and raise an error if multiple pages have the same title (maybe I have overlooked it though or this scenario is not possible)

Reply:
That's a good point. It's hard cause the path and the title can be different, but also provide the same thing. We would need to detect uniqueness of the hashes. I'll talk to @sfc-gh-jcarroll

Reply:
Addressed this.